store/src/test/java/com/dropbox/android/external/store4/impl/FlowTestExt.kt

```diff
@@ -15,19 +15,56 @@
  */
 package com.dropbox.android.external.store4.impl
 
+import com.google.common.truth.FailureMetadata
+import com.google.common.truth.Subject
+import com.google.common.truth.Truth
 import com.google.common.truth.Truth.assertThat
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.take
-import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.TestCoroutineScope
 
-/**
- * Asserts only the [expected] items by just taking that many from the stream
- *
- * Use this when Pipeline has an infinite part (e.g. Persister or a never ending fetcher)
- */
-@ExperimentalCoroutinesApi
-suspend fun <T> Flow<T>.assertItems(vararg expected: T) {
-    assertThat(this.take(expected.size).toList())
-            .isEqualTo(expected.toList())
+@UseExperimental(ExperimentalCoroutinesApi::class)
+internal fun <T> TestCoroutineScope.assertThat(flow: Flow<T>): FlowSubject<T> {
+    return Truth.assertAbout(FlowSubject.Factory<T>(this)).that(flow)
+}
+
+@UseExperimental(ExperimentalCoroutinesApi::class)
+internal class FlowSubject<T> constructor(
+    failureMetadata: FailureMetadata,
+    private val testCoroutineScope: TestCoroutineScope,
+    private val actual: Flow<T>
+) : Subject(failureMetadata, actual) {
+    /**
+     * Takes all items in the flow that are available by collecting on it as long as there are
+     * active jobs in the given [TestCoroutineScope].
+     *
+     * It ensures all expected items are dispatched as well as no additional unexpected items are
+     * dispatched.
+     */
+    suspend fun emitsExactly(vararg expected: T) {
+        val collectedSoFar = mutableListOf<T>()
+        val collectJob = testCoroutineScope.launch {
+            actual.collect {
+                collectedSoFar.add(it)
+                assertThat(collectedSoFar.size).isAtMost(expected.size)
```

Comment:
It looks like this exception will never propagate to the consumer, instead they'll always get the exception thrown by `assertThat(collectedSoFar).isEqualTo(expected.toList())` at the end.

This should work:
```
...
testCoroutineScope.advanceUntilIdle()
if (collectJob.isCancelled) {
    throw collectJob.getCancellationException().cause!!
} else {
    collectJob.cancel()
}
assertThat(collectedSoFar).isEqualTo(expected.toList())

Reply:
actually thats not what i'm seeing.

```
@Test
    fun moreThanExpectedItems() = testScope.runBlockingTest {
        val src = flowOf(1, 2, 3, 4)
        val exception = try {
            assertThat(src).emitsExactly(1, 2, 3)
            null
        } catch (error: AssertionError) {
            error
        }
        assertThat(exception).isInstanceOf(AssertionError::class.java)
        assertThat(exception).hasMessageThat().contains("unexpected (1): 4")
    }
```
fails with: 

```
expected to be at most: 3
but was               : 4
```

I'll update that message though as it is useless without telling what was the list in the first place.

Reply:
oh i see what you mean, because of launch, run blocking picks up that exception even though return value is the inner exception :man_facepalming: 
replaced w/ async to fix it, thanks!

Reply:
TIL how `launch` and `async` treat uncaught exceptions differently. Thanks!